provider/aws/aws_test.go

```diff
@@ -1100,6 +1100,50 @@ func TestAWSSuitableZones(t *testing.T) {
 	}
 }
 
+func TestAWSHealthTargetAnnotation(tt *testing.T) {
+	comparator := func(name, previous, current string) bool {
+		return previous == current
+	}
+	for _, test := range []struct {
+		name               string
+		current            *endpoint.Endpoint
+		desired            *endpoint.Endpoint
+		propertyComparator func(name, previous, current string) bool
+		shouldUpdate       bool
+	}{
+		{
+			name: "skip AWS target health",
+			current: &endpoint.Endpoint{
+				RecordType: "A",
+				DNSName:    "foo.com",
+				ProviderSpecific: []endpoint.ProviderSpecificProperty{
+					{Name: "aws/evaluate-target-health", Value: "true"},
+				},
+			},
+			desired: &endpoint.Endpoint{
+				DNSName:    "foo.com",
+				RecordType: "A",
+				ProviderSpecific: []endpoint.ProviderSpecificProperty{
+					{Name: "aws/evaluate-target-health", Value: "false"},
+				},
+			},
+			propertyComparator: comparator,
+			shouldUpdate:       false,
+		},
+	} {
+		tt.Run(test.name, func(t *testing.T) {
+			provider := &AWSProvider{}
+			plan := &plan.Plan{
+				Current:            []*endpoint.Endpoint{test.current},
+				Desired:            []*endpoint.Endpoint{test.desired},
+				PropertyComparator: provider.PropertyValuesEqual,
```

Comment:
Should this use `test.propertyComparator`?

Reply:
PropertyComparator should be removed as in #1849 because it can be implemented with AdjustEndpoint. There's no need for duplication in logic and PropertyComparator is used only by CloudFlare provider which I currently own.

Reply:
+1, we can remove it.

Reply:
EDIT: I'm not so sure anymore. What about https://github.com/kubernetes-sigs/external-dns/blob/904bf839eeaa96603fa62e9abc595f1992d52fe4/provider/aws/aws.go#L209-L211 ? I think we need the two functionality: one is about doing side effects on the records in a provider specific way because the provider "know better", the other is implementing a comparison. In #1849 those were mixed (like in https://github.com/kubernetes-sigs/external-dns/pull/1849/files#diff-efafdf56d71f1a3f19d003c8bad9fbadb8ee34e873cfa41362ca880e9f7c5fd3R382-R384 ), but I'm not thinking that we should probably keep the distinction.

Reply:
The code you are linking can be implemented with AdjustEndpoints by removing hard-setting value of `aws/evaluate-target-health` custom property to what is currently defined. I think implementing this would require changing signature of AdjustEndpoints to `AdjustEndpoints(current []*endpoint.Endpoint, desired []*endpoint.Endpoint) []*endpoint.Endpoint` where the result are adjusted desired endpoints. Actually it would also make sense to change name of `AdjustEndpoints` to `AdjustDesiredEndpoints` or something like this.

Reply:
Yet somehow it doesn't feel right either. I'm thinking that it could be more reasonable to change `AdjustEndpoints` to `ChangedEndpoints(current []*endpoint.Endpoint, desired []*endpoint.Endpoint) []*endpoint.Endpoint`, which would take current and desired endpoints, and return only these that should be changed. The default implementation in BaseProvider would be something like code in `plan.Calculate()`, and providers could override it (also call implementation of BaseProvider).

Reply:
Effectively this would render plan.go obsolete and would make planning (calculating needed changes) provider-specific with default implementation in BaseProvider

Reply:
Thank you for the detailed comments @sheerun , those are much appreciated. I generally agree with you, the idea that the plan can be generic was probably naive. I'll do more thinking on that during day time and update with a comment. The reason why I opened the PR in this state is exactly because I wanted to have those discussions and it's great if we can make ExternalDNS better structured, easier to maintain and to extend.

Reply:
If I were to redesign external-dns from scratch there would be per-provider `Change` type with common interface for logging, where each Change maps maps 1:1 to API calls of provider (for easy testing). Also sources would have streaming interface (i.e. they would only return changed and removed endpoints). The main loop would be something like:

```go
config := ReadConfig()
source, providers := InstantiateSourceAndProviders(config)

for { 
	// source.GetChangedEndpoints blocks until there is an update or context is cancelled
	// endpoints are just changed endpoints, if endpoint is removed then endpoint.targets is empty
	endpoints := source.GetChangedEndpoints(ctx)
	for _, provider := range providers {
		// provider returns needed changes, possibly updating internal "current" state
		// probably ignoring most of changes because domain is not handled by provider
		changes := provider.GetChanges(ctx, endpoints)
		logger.LogChanges(changes)
		provider.ApplyChanges(ctx, changes)
	}
}
```

This eliminates a lot of complexity, configuration, and reduces time spent on unnecessary actions (e.g. fetching dns records of domains that are not subject of changes, re-calculating needed changes). But it's probably too late for per-provider changes, but a version of above could be implemented now is:

1. Make all sources to return only changed endpoints (easy)
2. Make `provider.Records()` private for each provider (or just don't call it anywhere outside). This method is no longer necessary to be implemented.
3. Introduce `provider.GetChanges(endpoints)` that returns current `Changes` type (with common implementation in BaseProvider.GetChanges)
4. Get rid of plan.go and managing "current" records outside of providers